### PR TITLE
Implement `complex` intrinsic function

### PR DESCRIPTION
**Pull request by namannimmo10**
expr2.py contains:
```python
def main():
    c: c128
    c = complex()
    c = complex(3.4)
    c = complex(5., -4.3)
```
ASR:
```bash
(TranslationUnit
   (SymbolTable
      1
      {
         main:
            (Subroutine
               (SymbolTable
                  2
                  {
                     c:
                        (Variable
                           2
                           c
                           Local () ()
                           Default
                           (Complex 8 [])
                           Source
                           Public
                           Required .false.)
                  })
               main [] [
               (=
                  (Var 2 c)
                  (ConstantComplex 0.000000 0.000000
                     (Complex 8 [])) ())
               (=
                  (Var 2 c)
                  (ConstantComplex 3.400000 0.000000
                     (Complex 8 [])) ())
               (=
                  (Var 2 c)
                  (ConstantComplex 5.000000 -4.300000
                     (Complex 8 [])) ())]
               Source
               Public
               Implementation () .false. .false.)
      })
   [])
```

### CODE REVIEW

**Comment by certik** [comment]
Do you mind adding the test case into `doconcurrentloop_01.py`? That way it will get tested by the CI.

File: src/lfortran/semantics/python_ast_to_asr.cpp [hunk 1]
```diff
@@ -1212,6 +1212,29 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
+            } else if (n_args == 2) {
+                ASR::expr_t* expr1 = args[0];
+                ASR::expr_t* expr2 = args[1];
+                double c1 = ASR::down_cast<ASR::ConstantReal_t>(ASRUtils::expr_value(expr1))->m_r;
+                double c2 = ASR::down_cast<ASR::ConstantReal_t>(ASRUtils::expr_value(expr2))->m_r;
+                tmp = ASR::make_ConstantComplex_t(al, x.base.base.loc, c1, c2, type);
+                return;
+            } else {
+                throw SemanticError("complex() must have at most two real arguments", x.base.base.loc);
+            }
```
Comment (czgdp1807) [inline src/lfortran/semantics/python_ast_to_asr.cpp:1237]:
```suggestion
            if( n_args > 2 || n_args < 0 ) { // n_args shouldn't be less than 0 but added this check for safety
                throw SemanticError("complex() must have at most two real arguments", x.base.base.loc);
            }

            double c1 = 0.0, c2 = 0.0; // Default if n_args = 0
            if (n_args >= 1) { // Handles both n_args = 1 and n_args = 2
                c1 = ASR::down_cast<ASR::ConstantReal_t>(ASRUtils::expr_value(args[0]))->m_r;
            }

            if (n_args == 2) { // Extracts imaginary component if n_args = 2
                double c2 = ASR::down_cast<ASR::ConstantReal_t>(ASRUtils::expr_value(args[1]))->m_r;
            }

            tmp = ASR::make_ConstantComplex_t(al, x.base.base.loc, c1, c2, type);
```

**Comment by czgdp1807** [comment]
Was going through the PRs and left the above suggestion.

**Comment by czgdp1807** [comment]
Just a minor question, Is it legal to do something like the following,
```
r = 1
i = 2
c = complex(r, i)
```
In words, I mean passing variables as arguments to `complex`?

**Comment by Thirumalai-Shaktivel** [comment]
This is not mentioned in the [Docs](https://docs.python.org/3/library/functions.html#complex), But the interactive python shell accepts it!
```
Python 3.8.5 (default, Sep  4 2020, 07:30:14) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> r = 1
>>> i = 2
>>> c = complex(r, i)
>>> c
(1+2j)
>>> type(c)
<class 'complex'>
```
which means it is legal to pass the variables as parameters.

File: src/lfortran/semantics/python_ast_to_asr.cpp [hunk 2]
```diff
@@ -1212,6 +1212,29 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
             } else {
                 throw SemanticError("chr() must have one integer argument", x.base.base.loc);
             }
+        } else if (call_name == "complex") {
+            LFORTRAN_ASSERT(ASRUtils::all_args_evaluated(args));
```
Comment (czgdp1807) [inline src/lfortran/semantics/python_ast_to_asr.cpp:1216]: I think that if variables are allowed then this check should be removed as this accepts only those arguments which are constants or can be evaluated to a constant at compile time.

Comment (Thirumalai-Shaktivel) [reply src/lfortran/semantics/python_ast_to_asr.cpp:1216]: Yup, I agree with you! 
But, I think naman is planning to support only constant values for now. (I don't know)

Comment (Thirumalai-Shaktivel) [reply src/lfortran/semantics/python_ast_to_asr.cpp:1216]: In that case, we can add a comment saying "Only constant numeric values as parameter are supported for now". Iteratively we can support for both strings(`complex('1+2j')` or `complex('1')`) and variables (`complex(x)`)

Comment (namannimmo10) [reply src/lfortran/semantics/python_ast_to_asr.cpp:1168]: Thanks.  

> I think that if variables are allowed then this check should be removed as this accepts only those arguments which are constants or can be evaluated to a constant at compile time.

Yes, this should be done. I guess we also need to modify other intrinsic functions in this file (Python AST->ASR), like `chr` and `ord`, in order to make support for variables.

**Review by certik — APPROVED** [review]
I think this is fine. This is a compile time function, which can be used to construct complex numbers. Once we have a runtime library working, we can add a runtime implementation.

**Comment by Smit-create** [comment]
The CI passes for this PR. I'm just re-running the jobs after we have a few updates in CI to verify if that fails here.

**Comment by Thirumalai-Shaktivel** [comment]
This PR is not updated (I mean, not pushed) since PR #33  is merged, that's why it uses the old CI...

**Comment by certik** [comment]
What is the status of this PR --- can it go in?

**Comment by namannimmo10** [comment]
I think this is ready once the build is a success. We can extend this later.

File: doconcurrentloop_01.py [hunk 3]
```diff
@@ -1,3 +1,21 @@
...
+    # constant real or int as args
+    c = complex()
+    c = complex(3.4)
+    c = complex(5., 4.3)
+    c = complex(1)
+    c1 = complex(3, 4)
+    c2 = complex(2, 4.5)
+    c3 = complex(3., 4.)
+    b = c1 != c2
+    b = c1 == c3
```
Comment (certik) [inline doconcurrentloop_01.py:17]: Add these changes to `lpython_tests.py`, after #37 is merged. Then it should work.

**Pull request merged**